### PR TITLE
Added eea_getTransactionCount Json Rpc

### PR DESCRIPTION
**Pull request by rain-on**
Exposers a Json Rpc which allows "legacy" users to query for the
nonce of a given address, in a group of private users.

### CODE REVIEW

File: ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/EeaGetTransactionCount.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.collect.Lists;
+import org.apache.logging.log4j.Logger;
+import org.bouncycastle.util.Arrays;
+
+public class EeaGetTransactionCount implements JsonRpcMethod {
+
+  public static class EeaGetTransactionCountJsonBody {
+
+    final String privateFrom;
```
Comment (CjHare) [inline ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/EeaGetTransactionCount.java]: Couldn't these be `final`?

File: ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/EeaGetTransactionCount.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.collect.Lists;
+import org.apache.logging.log4j.Logger;
+import org.bouncycastle.util.Arrays;
+
+public class EeaGetTransactionCount implements JsonRpcMethod {
+
+  public static class EeaGetTransactionCountJsonBody {
```
Comment (CjHare) [inline ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/EeaGetTransactionCount.java]: Pattern used in the JsonRpc package has the Json POJO split out into their classes and suffixed with `Parameters` e.g. `tech.pegasys.pantheon.ethereum.jsonrpc.internal.parameters`, something worth considering?

File: ethereum/jsonrpc/src/test/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/EeaGetTransactionCountTest.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+import tech.pegasys.pantheon.ethereum.privacy.PrivateTransactionHandler;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.junit.Before;
+import org.junit.Test;
+
+public class EeaGetTransactionCountTest {
+  final Address address = Address.fromHexString("55");
```
Comment (CjHare) [inline ethereum/jsonrpc/src/test/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/EeaGetTransactionCountTest.java]: `private` modifier?

File: ethereum/jsonrpc/src/test/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/EeaGetTransactionCountTest.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        };
+
+    when(enclave.findPrivacyGroup(any())).thenReturn(returnedGroups);
+    when(privateTransactionHandler.getSenderNonce(address, "Group1")).thenReturn(8L);
+
+    final EeaGetTransactionCount method =
+        new EeaGetTransactionCount(new JsonRpcParameter(), enclave, privateTransactionHandler);
+
+    final JsonRpcResponse response = method.response(request);
+    assertThat(response).isInstanceOf(JsonRpcErrorResponse.class);
```
Comment (CjHare) [inline ethereum/jsonrpc/src/test/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/EeaGetTransactionCountTest.java]: Are you expecting a specific error message here?

File: ethereum/jsonrpc/src/test/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/EeaGetTransactionCountTest.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+  }
+
+  @Test
+  public void enclaveThrowingAnExceptionResultsInErrorResponse() throws Exception {
+    when(enclave.findPrivacyGroup(any())).thenThrow(new Exception("Enclave Failed"));
+    final EeaGetTransactionCount method =
+        new EeaGetTransactionCount(new JsonRpcParameter(), enclave, privateTransactionHandler);
+
+    final JsonRpcResponse response = method.response(request);
+    assertThat(response).isInstanceOf(JsonRpcErrorResponse.class);
```
Comment (CjHare) [inline ethereum/jsonrpc/src/test/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/EeaGetTransactionCountTest.java]: ditto, is any JsonRpcErrorResponse acceptable? or do you expecting a specific message?

File: ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/EeaGetTransactionCount.java [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+import java.util.List;
+import java.util.stream.Collectors;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.collect.Lists;
+import org.apache.logging.log4j.Logger;
+import org.bouncycastle.util.Arrays;
+
+public class EeaGetTransactionCount implements JsonRpcMethod {
```
Comment (jframe) [inline ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/EeaGetTransactionCount.java]: Think this class should be name PrivGetLegacyTransactionCount

File: ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/EeaGetTransactionCount.java [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+      LOG.error("Failed to fetch group from Enclave with error " + e.getMessage());
+      LOG.error(e);
+      return new JsonRpcErrorResponse(request.getId(), JsonRpcError.FIND_PRIVACY_GROUP_ERROR);
+    }
+  }
+
+  private long determineNonce(final EeaGetTransactionCountJsonBody getTransactionCount)
+      throws Exception {
+
+    final String[] groupMembers =
```
Comment (jframe) [inline ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/EeaGetTransactionCount.java]: Perhaps there should be a method on the PrivateTransactionHandler to get the legacy privacy group?

File: ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/EeaGetTransactionCount.java (hunk 1 above)
Comment (rain-on) [reply ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/EeaGetTransactionCount.java]: Private? yes.

File: ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/EeaGetTransactionCount.java (hunk 2 above)
Comment (rain-on) [reply ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/EeaGetTransactionCount.java]: done.

File: ethereum/jsonrpc/src/test/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/EeaGetTransactionCountTest.java (hunk 3 above)
Comment (rain-on) [reply ethereum/jsonrpc/src/test/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/EeaGetTransactionCountTest.java]: Done.

File: ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/EeaGetTransactionCount.java (hunk 6 above)
Comment (rain-on) [reply ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/EeaGetTransactionCount.java]: Talked to Madeline last night - given this capability is really required of Legacy Privacy - it belongs in the Eea namespace, and thus, the spec will need updating.

Comment (rain-on) [reply ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/EeaGetTransactionCount.java]: More to the point - priv_ namespace is geared toward the new Privacy Group - so having "priv" and "legacy" in a name is not a valid combination.

File: ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/EeaGetTransactionCount.java (hunk 7 above)
Comment (rain-on) [reply ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/EeaGetTransactionCount.java]: I couldn't decide if this is something a PrivateTransactionHandler should know about ...

File: ethereum/jsonrpc/src/test/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/EeaGetTransactionCountTest.java (hunk 5 above)
Comment (rain-on) [reply ethereum/jsonrpc/src/test/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/EeaGetTransactionCountTest.java]: Done.

File: ethereum/jsonrpc/src/test/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/EeaGetTransactionCountTest.java (hunk 4 above)
Comment (rain-on) [reply ethereum/jsonrpc/src/test/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/EeaGetTransactionCountTest.java]: Done.

File: ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/LegacyNonceProvider.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+import tech.pegasys.pantheon.ethereum.privacy.PrivateTransactionHandler;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import com.google.common.collect.Lists;
+import org.apache.logging.log4j.Logger;
+import org.bouncycastle.util.Arrays;
+
+public class LegacyNonceProvider {
```
Comment (CjHare) [inline ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/LegacyNonceProvider.java]: Why does this have the `Legacy` prefix, rather than `Eea`?

File: ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/LegacyNonceProvider.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+  private final Enclave enclave;
+  private final PrivateTransactionHandler privateTransactionHandler;
+
+  public LegacyNonceProvider(
+      final Enclave enclave, final PrivateTransactionHandler privateTransactionHandler) {
+    this.enclave = enclave;
+    this.privateTransactionHandler = privateTransactionHandler;
+  }
+
+  public long determineNonce(
```
Comment (CjHare) [inline ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/LegacyNonceProvider.java]: Is there a more narrow exception to throw instead of `Exception`?

Comment (rain-on) [reply ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/LegacyNonceProvider.java]: alas - this is an effect of Encalve throwing Exception ... this behaviour is mimic'd from other Priv Json RPC ...

File: ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/LegacyNonceProvider.java (hunk 8 above)
Comment (rain-on) [reply ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/LegacyNonceProvider.java]: Because "Legacy" is the default terminology for things which work with EEA privacy ... (i.e. internally the Privacy is deemed LEGACY or PANTHEON.

File: ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/EeaGetTransactionCount.java [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+
+    final Address address = parameters.required(request.getParams(), 0, Address.class);
+    final String privateFrom = parameters.required(request.getParams(), 1, String.class);
+    final String[] privateFor = parameters.required(request.getParams(), 2, String[].class);
+
+    try {
+      final long nonce = nonceProvider.determineNonce(privateFrom, privateFor, address);
+      return new JsonRpcSuccessResponse(request.getId(), Quantity.create(nonce));
+    } catch (final Exception e) {
+      LOG.error("Failed to fetch group from Enclave with error " + e.getMessage());
```
[the remark this answers is not in the document]
Comment (rain-on) [reply ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/EeaGetTransactionCount.java]: Done.

File: ethereum/core/src/test/java/tech/pegasys/pantheon/ethereum/mainnet/precompiles/privacy/PrivacyPrecompiledContractTest.java [hunk 11]
```diff
@@ -107,7 +107,7 @@ private PrivateTransactionProcessor mockPrivateTxProcessor() {
 
   private Enclave brokenMockEnclave() throws Exception {
     Enclave mockEnclave = mock(Enclave.class);
```
Comment (jframe) [inline ethereum/core/src/test/java/tech/pegasys/pantheon/ethereum/mainnet/precompiles/privacy/PrivacyPrecompiledContractTest.java]: throws Exception shouldn't be needed now it's throwing the runtime EnclaveException

File: ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/LegacyNonceProvider.java [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+      final String privateFrom, final String[] privateFor, final Address address) {
```
Comment (jframe) [inline ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/LegacyNonceProvider.java]: nit: Think it would have been nice if the Enclave provided a findEeaPrivacyGroup(privateFor, privateFrom) method. Probably not part of this PR though

Comment (rain-on) [reply ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/LegacyNonceProvider.java]: Yup - this should probably be Orion.

File: ethereum/core/src/test/java/tech/pegasys/pantheon/ethereum/mainnet/precompiles/privacy/PrivacyPrecompiledContractTest.java (hunk 11 above)
Comment (rain-on) [reply ethereum/core/src/test/java/tech/pegasys/pantheon/ethereum/mainnet/precompiles/privacy/PrivacyPrecompiledContractTest.java]: done.

File: enclave/src/main/java/tech/pegasys/pantheon/enclave/Enclave.java [hunk 13]
```diff
@@ -60,47 +63,64 @@ public boolean upCheck() throws IOException {
...
-      final MediaType mediaType, final Object content, final String endpoint) throws Exception {
-    final RequestBody body =
-        RequestBody.create(mediaType, objectMapper.writeValueAsString(content));
+      final MediaType mediaType, final Object content, final String endpoint) {
+    final RequestBody body;
+    try {
+      body = RequestBody.create(mediaType, objectMapper.writeValueAsString(content));
+    } catch (final JsonProcessingException e) {
+      final String errorMessage = "Failed to serialise request to json body.";
+      LOG.error(errorMessage);
```
[the remark this answers is not in the document]
Comment (rain-on) [reply enclave/src/main/java/tech/pegasys/pantheon/enclave/Enclave.java]: done

File: ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/EeaGetTransactionCount.java [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+    final Address address = parameters.required(request.getParams(), 0, Address.class);
+    final String privateFrom = parameters.required(request.getParams(), 1, String.class);
+    final String[] privateFor = parameters.required(request.getParams(), 2, String[].class);
+
+    try {
+      final long nonce = nonceProvider.determineNonce(privateFrom, privateFor, address);
+      return new JsonRpcSuccessResponse(request.getId(), Quantity.create(nonce));
+    } catch (final Exception e) {
+      LOG.error(e.getMessage(), e);
+      return new JsonRpcErrorResponse(request.getId(), JsonRpcError.FIND_PRIVACY_GROUP_ERROR);
```
Comment (jframe) [inline ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/EeaGetTransactionCount.java]: this looks like the wrong error type

Comment (rain-on) [reply ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/EeaGetTransactionCount.java]: done.

File: ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/EeaPrivateNonceProvider.java [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+    final List<PrivacyGroup> legacyGroups =
+        matchingGroups.stream()
+            .filter(group -> group.getType() == Type.LEGACY)
+            .collect(Collectors.toList());
+
+    if (legacyGroups.size() != 1) {
+      final String errorMessage =
+          String.format(
+              "Found invalid number of privacy groups (%d), expected 1.", legacyGroups.size());
+      LOG.error(errorMessage);
```
Comment (jframe) [inline ethereum/jsonrpc/src/main/java/tech/pegasys/pantheon/ethereum/jsonrpc/internal/methods/privacy/eea/EeaPrivateNonceProvider.java]: no need to log and throw, perhaps just throw with the error message. Will be logged in rpc layer
